### PR TITLE
fix(neighbors): derive JAX search radius from realized grids

### LINKED ISSUE #128: JAX batched cell list can miss neighbors with static max_total_cells
### Environment

- nvalchemi-toolkit-ops 0.4.0
- JAX 0.11.0
- Warp 1.15.0
- CUDA GPU

### Problem

`batch_build_cell_list` requires `max_total_cells` inside `jax.jit`, but when that argument is supplied it initializes `neighbor_search_radius` to ones instead of deriving it from the generated `cells_per_dimension` and cell face distances. The returned radius can therefore be too small and `batch_cell_list` / `batch_query_cell_list` silently omit valid neighbors.

For an identity cell, cutoff 0.3, and non-periodic random points, the builder returns `cells_per_dimension = [6, 6, 6]` but `neighbor_search_radius = [1, 1, 1]`. Since each bin is 1/6 wide, a radius of 2 is required for a 0.3 cutoff. In a 128-query / 128-key test using the public `target_indices` path, a brute-force reference found 1329 bipartite pairs while the cell list returned 1077.

### Workaround / expected behavior

Calling public `batch_build_cell_list`, replacing the radius with

```python
face = 1 / jnp.linalg.norm(jnp.linalg.inv(cell).transpose(0, 2, 1), axis=2)
radius = jnp.ceil(jnp.nextafter(cutoff * cells_per_dimension / face, jnp.inf)).astype(jnp.int32)
radius = jnp.where((cells_per_dimension == 1) & ~pbc, 0, radius)
```

and then calling public `batch_query_cell_list` restores exact edge-set equality with brute force and kUPS for all eight 3D PBC masks.

The `max_total_cells is None` branch calls the estimator and receives a derived radius, but that branch is documented as not JIT-compatible. The static-capacity branch should derive the radius after constructing the actual bin counts as well.

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Fix JAX cell-list neighbor queries so automatically determined search radii are derived from the realized cell grid instead of placeholder or pre-build values. Align JAX adaptive cell promotion with the Warp build kernels, provide exact batched estimator metadata for selected capacity policies, and reject fused Warp graph calls that cannot determine search metadata before querying.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes https://github.com/NVIDIA/nvalchemi-toolkit-ops/issues/128.

## Changes Made

- Add a shared JAX helper that derives per-axis neighbor search radii from cell face distances, realized cell counts, cutoff, and PBC flags.
- Recompute automatic radii after single-system and batched cell-list construction while preserving explicit caller-provided radii.
- Match Warp's repeated-doubling adaptive promotion and require an explicit radius for fused Warp graph queries with static cell capacity.
- Keep volume capacity as the default while adding opt-in geometry capacity for the JAX batch estimator; estimator grids and radii now match construction for the returned capacity.
- Reserve one constructible cell per empty batch system and avoid a duplicate construct launch in automatic builds.
- Add JIT, static-capacity, adaptive-promotion, compact-target, PBC-mask, naive-parity, and fused-graph regression coverage.
- Add capacity-strategy coverage for exact grid/radius values, estimator/build parity across dtypes and cell shapes, invalid strategies, empty batches, and automatic-build construct dispatch.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

- JAX single-system and batched cell-list suites: 115 passed and 8 optional `vesin` skips.
- Targeted Torch batch cell-list estimator regressions: 8 passed and 4 slow compile cases skipped.
- Full `make pytest`: 2,412 tests passed and 437 skipped before collection stopped because optional `torchpme` was unavailable for the Torch Ewald tests.
- `make lint`: passed.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

The fused Warp graph path requires an explicit `neighbor_search_radius` when `max_total_cells` is supplied because the query launch needs concrete search metadata before build output is available.

The batch estimator defaults to volume capacity. Use `capacity_strategy="geometry"` to estimate from promoted per-axis grid geometry, then pass the returned capacity to `batch_build_cell_list`.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/jax/neighbors/batch_cell_list.py [hunk 1]
```diff
@@ -1007,10 +1006,14 @@ def estimate_batch_cell_list_sizes(
...
     cells_per_dimension = jnp.stack(cells_per_dim_list, axis=0)
-    neighbor_search_radius = jnp.stack(search_radius_list, axis=0)
+    neighbor_search_radius = _derive_neighbor_search_radius(
+        cell,
+        _pbc_bool,
+        cutoff,
+        cells_per_dimension,
+    )
 
     return max_total_cells, cells_per_dimension, neighbor_search_radius
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/batch_cell_list.py]: Fixed in 23dc0ab4e8708ee56c1a62623c70e18f3f252867.

File: nvalchemiops/jax/neighbors/batch_cell_list.py [hunk 2]
```diff
@@ -387,6 +390,86 @@ def _normalize_batch_cell_pbc(
...
+            )
+
+        system_capacity = max(num_cells_est, 8)
+        if capacity_strategy == "geometry":
+            max_total_cells = max(max_total_cells, system_capacity)
+        else:
+            max_total_cells += system_capacity
+
+    if capacity_strategy == "geometry":
+        return max(max_total_cells * num_systems, num_systems)
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/batch_cell_list.py]: Style, but could just do

```python
if capacity_strategy == "geometry":
    max_total_cells *= num_systems
return max(max_total_cells, num_systems)
```

File: nvalchemiops/jax/neighbors/batch_cell_list.py [hunk 3]
```diff
@@ -387,6 +390,86 @@ def _normalize_batch_cell_pbc(
...
+
+
+def _construct_batch_cells_per_dimension(
+    cell: jax.Array,
+    pbc: jax.Array,
+    cutoff: float,
+    max_total_cells: int,
+) -> jax.Array:
+    """Run the authoritative construct kernel and return realized bins."""
+    num_systems = cell.shape[0]
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/batch_cell_list.py:456]: Here it would be good to check and raise

```python
if max_total_cells < num_systems: raise ValueError
```

or do it in `estimate_batch_cell_list_sizes`

File: nvalchemiops/jax/neighbors/batch_cell_list.py [hunk 4]
```diff
@@ -1117,15 +1213,19 @@ def batch_build_cell_list(
...
-        max_total_cells, cells_per_dim_est, neighbor_search_radius = (
-            estimate_batch_cell_list_sizes(
-                positions, batch_ptr, batch_idx, cell, cutoff, pbc
-            )
+        max_total_cells = _estimate_batch_max_total_cells(
+            batch_ptr,
+            cell,
+            pbc_bool,
+            cutoff,
+            1.5,
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/batch_cell_list.py]: Why is this hardcoded?

File: nvalchemiops/jax/neighbors/batch_cell_list.py [hunk 5]
```diff
@@ -387,6 +390,86 @@ def _normalize_batch_cell_pbc(
...
+
+    num_systems = batch_ptr.shape[0] - 1
+    promoted_cells_per_dimension = (
+        _derive_promoted_cells_per_dimension(cell, pbc, cutoff)
+        if capacity_strategy == "geometry"
+        else None
+    )
+    max_total_cells = 0
+
+    for sys_idx in range(num_systems):
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/batch_cell_list.py:417]: For maintainability, dropping some inline comments to indicate what is happening in the loop and why would be good

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,16 @@
 # Changelog
 
+## Unreleased
```
Comment (laserkelvin) [inline CHANGELOG.md:3]: I think maybe do 0.4.1 - Unreleased

**Review by laserkelvin — COMMENTED** [review]
Minor things

**Comment by laserkelvin** [comment]
/ok to test [3b50d67](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/129/commits/3b50d67010656e8372c3effe0c02578ce22dd83f)

File: CHANGELOG.md (hunk 6 above)
Comment (zubatyuk) [reply CHANGELOG.md:3]: Will do follow-up in RC branch.

File: nvalchemiops/jax/neighbors/batch_cell_list.py (hunk 4 above)
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/batch_cell_list.py]: Fixed in bc2efedde068bae5bf2d3d83213f78bf62c21b11.

File: nvalchemiops/jax/neighbors/batch_cell_list.py (hunk 3 above)
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/batch_cell_list.py:456]: Fixed in c950235b5952e7dc2d3c20249119bcb5ad7d49ab.

File: nvalchemiops/jax/neighbors/batch_cell_list.py (hunk 2 above)
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/batch_cell_list.py]: Fixed in 8f750ef95c927619a57e7d44f0848d42779a860b.

File: nvalchemiops/jax/neighbors/batch_cell_list.py (hunk 5 above)
Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/batch_cell_list.py:417]: Fixed in 09de673f762f3c0c8b928854af4c1a99d17a7bc3.

**Comment by laserkelvin** [comment]
/ok to test [09de673](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/129/commits/09de673f762f3c0c8b928854af4c1a99d17a7bc3)

**Review by laserkelvin — APPROVED** [review]

**Comment by laserkelvin** [comment]
Force merging - merge queue borked for some reason but the status checks all passed.
